modules/network/src/main/java/org/apache/ignite/network/scalecube/ScaleCubeDirectMarshallerTransport.java

```diff
@@ -0,0 +1,243 @@
+            return Address.create(address.getHostAddress(), port);
+    }
+
+    /**
+     * Cleanup resources on stop.
+     *
+     * @return A mono, that resolves when the stop operation is finished.
+     */
+    private Mono<Void> doStop() {
+        return Mono.defer(() -> {
+            LOG.info("[{0}][doStop] Stopping", address);
+
+            // Complete incoming messages observable
+            sink.complete();
+
+            LOG.info("[{0}][doStop] Stopped", address);
+            return Mono.empty();
+        });
+    }
+
+    /** {@inheritDoc} */
+    @Override public Address address() {
+        return address;
+    }
+
+    /** {@inheritDoc} */
+    @Override public Mono<Transport> start() {
+        return Mono.just(this);
+    }
+
+    /** {@inheritDoc} */
+    @Override public Mono<Void> stop() {
+        return doStop();
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isStopped() {
+        return onStop.isDisposed();
+    }
+
+    /** {@inheritDoc} */
+    @Override public Mono<Void> send(Address address, Message message) {
+        var addr = InetSocketAddress.createUnresolved(address.host(), address.port());
+        return Mono.defer(() -> Mono.fromFuture(connectionManager.channel(addr)))
+            .flatMap(client -> {
+                client.send(fromMessage(message));
+                return Mono.empty();
+            });
+    }
+
+    /**
+     * Handles new network messages from {@link #connectionManager}.
+     *
+     * @param source Message source.
+     * @param msg Network message.
+     */
+    private void onMessage(InetSocketAddress source, NetworkMessage msg) {
+        Message message = fromNetworkMessage(msg);
+
+        if (message != null)
+            sink.next(message);
+    }
+
+    /**
+     * Wrap ScaleCube {@link Message} with {@link NetworkMessage}.
+     *
+     * @param message ScaleCube message.
+     * @return Netowork message that wraps ScaleCube message.
+     * @throws IgniteInternalException If failed to write message to ObjectOutputStream.
+     */
+    private NetworkMessage fromMessage(Message message) throws IgniteInternalException {
+        Object dataObj = message.data();
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+
+        try(ObjectOutputStream oos = new ObjectOutputStream(stream)) {
+            oos.writeObject(dataObj);
+        }
+        catch (IOException e) {
+            throw new IgniteInternalException(e);
+        }
+
+        return new ScaleCubeMessage(stream.toByteArray(), message.headers());
+    }
+
+    /**
+     * Unwrap ScaleCube {@link Message} from {@link NetworkMessage}.
+     *
+     * @param networkMessage Network message.
+     * @return ScaleCube message.
+     * @throws IgniteInternalException If failed to read ScaleCube message byte array.
+     */
+    private Message fromNetworkMessage(NetworkMessage networkMessage) throws IgniteInternalException {
+        if (networkMessage instanceof ScaleCubeMessage) {
```

Comment:
Should we assert this instead of checking?

Reply:
No, I don't think so. We receive every single message here